### PR TITLE
Various Improvements 

### PR DESCRIPTION
**Pull request by IsaiahKelly**
One important change here is the implementation of the `ComponentDataWrapper` class which allows us to attach `IComponentData` directly to `GameObjects` just like good-old `MonoBehaviours`. This eliminates the need to add entity components to prefabs through code! 

However, in order to make it work the `ComponentDataWrapper` class must have the same name as the script file and this must be different from the `IComponentData` class name. 

In the Unity official examples they just suffix "Component" to the end of the script file and `ComponentDataWrapper`, but I find this annoying. So I actually went with suffixing "Data" to the end of the `IComponentData` class instead. I think this works better because we're using a mix of old components and new _data_ components. So denoting the new pure data components from everything else seems very useful here.

I also moved some of the global game settings out to individual components so we can now actually setup unique values on each enemy prefab. Instead of having them all share the same exact health and damage values, which is less than ideal.

Finally I converted the new `GameSettings` class to a `ScriptableObject` singleton. Making it much easier to manage and now all changes made during play-mode persist.

Hope you find all this helpful!

### CODE REVIEW

**Comment by davidpol** [comment]
Thank you for taking the time to make these contributions! It is really appreciated.

The changes you mention definitely sound sensible, but I would be more tempted to go over them if they were smaller and more isolated. I do not really have the time right now to review thousands of changes, with many of them consisting on name or formatting changes that are not strictly needed.

**Comment by davidpol** [comment]
I was referring more to things like changing namespace names, the spacing format across the entire project (I guess due to auto-reformatting it with an IDE) or even the structure of the project itself (including the original Unity code); those are not really needed in the context of the improvements you mention but make it pretty hard for me to follow them because one ends up with thousands of changes across hundreds of files.

**Comment by IsaiahKelly** [comment]
Fair enough, but I think you might be exaggerating a wee bit about how drastic these changes are. Some of this also includes deleting hundreds of files and thousands of lines of code. Which I would think actually makes things easier to follow. In fact, the whole point of all my changes was to make things easier to understand and follow (at least for those studying the example).

I would also argue that being able to follow and understand every little change here doesn't really matter in a project this simple, but that's just my opinion! Sorry to inconvenience you with this, I was only trying to be helpful!

**Comment by davidpol** [comment]
Not an inconvenience at all! As mentioned earlier, I appreciate the improvements and would be happy to merge them. I just do not see how changing the naming and spacing standards of the entire project is relevant or necessary. There are thousands of little changes (e.g., changing code like GetComponent<EnemyHealth>() to GetComponent<EnemyHealth> () - note the added space before the parentheses) in the pull request, which simply get in the way of the actually useful ones.

Also, the ECS-specific code was in a separate folder in the first place, so it is not exactly hard to follow.

**Comment by IsaiahKelly** [comment]
Yeah I understand your issue here. No need to repeat yourself. However, even if I redid this PR without the formatting changes, it would still probably be too much for you. So it's probably best if I just work on my own examples instead.

Cheers!

**Comment by davidpol** [comment]
If you redid the PR without the formatting changes, I would be glad to merge it. That is kind of what I have been saying the whole time. In any case, you seem to have taken this way too personally, so I will just leave you be. Thank you again!

**Comment by davidpol** [comment]
If you redid the PR without the formatting changes, I would be glad to merge it. That is kind of what I have been saying all along. In any case, at this point in time, you seem to have taken this way too personally, so I will just leave you be. Thank you again!

**Comment by davidpol** [comment]
If you redid the PR without the unnecessary changes, I would be glad to merge it. That is kind of what I have been saying all along. In any case, at this point in time you seem to have taken this way too personally, so I will just leave you be. Thank you again!

**Comment by IsaiahKelly** [comment]
You misunderstand me. Again, I understand your issue here. I was just trying to say that I don't think any PR I submit would be free of "unnecessary changes" since I think my overall objective here is a little too grand for any PR you would accept. So better we just do our own things. Sorry. I wouldn't have made this PR in the first place if I had known this would be such an issue.

No hard feelings!

**Comment by davidpol** [comment]
If changing the naming and formatting conventions (including adding spurious spaces) to an existing codebase does not enter the "unnecessary changes" category for you, that will definitely be the best way to move forward. We will agree to disagree on that one. :-)

Thank you again for your interest and best of luck with your projects!

**Comment by davidpol** [comment]
If changing the naming and formatting conventions (including adding spurious spaces and mixing naming rules) together with changing the folder structure of an already established codebase does not enter the "unnecessary changes" category for you, that will definitely be the best way to move forward. We will just have to agree to disagree on that one. :-)

Thank you again for your interest and best of luck with your projects!
